dm/master/scheduler/scheduler.go

```diff
@@ -318,6 +331,124 @@ func (s *Scheduler) RemoveSourceCfg(source string) error {
 	return nil
 }
 
+// GetTaskCfg gets config of task
+func (s *Scheduler) GetTaskCfg(task string) string {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	cfg, ok := s.taskCfgs[task]
+	if !ok {
+		return ""
+	}
+	return cfg
+}
+
+// AddTaskCfg adds the config of task
+func (s *Scheduler) AddTaskCfg(cfg config.TaskConfig) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if !s.started {
+		return terror.ErrSchedulerNotStarted.Generate()
+	}
+
+	if len(cfg.Name) == 0 {
+		return nil
+	}
+
+	// 1. check whether exists.
+	if _, ok := s.taskCfgs[cfg.Name]; ok {
+		return terror.ErrSchedulerTaskExist.Generate(cfg.Name)
+	}
+
+	// 2. put the config into etcd.
+	clone := cfg
+	for _, mysqlInstance := range clone.MySQLInstances {
+		mysqlInstance.RemoveDuplicateCfg()
+	}
+
+	taskCfg := clone.String()
+	_, err := ha.PutTaskCfg(s.etcdCli, clone.Name, taskCfg)
+	if err != nil {
+		return err
+	}
+
+	// 3. record the config
+	s.taskCfgs[clone.Name] = taskCfg
+
+	return nil
+}
+
+// UpdateTaskCfg update the config of task
+func (s *Scheduler) UpdateTaskCfg(cfg config.TaskConfig) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if !s.started {
+		return terror.ErrSchedulerNotStarted.Generate()
+	}
+
+	if len(cfg.Name) == 0 {
+		return nil
+	}
+
+	// 1. check whether exists.
+	if _, ok := s.taskCfgs[cfg.Name]; !ok {
+		return terror.ErrSchedulerTaskNotExist.Generate(cfg.Name)
+	}
+
+	// 2. put the config into etcd.
+	clone := cfg
+	for _, mysqlInstance := range clone.MySQLInstances {
+		mysqlInstance.RemoveDuplicateCfg()
+	}
+
+	taskCfg := clone.String()
+	_, err := ha.PutTaskCfg(s.etcdCli, clone.Name, taskCfg)
+	if err != nil {
+		return err
+	}
+
+	// 3. record the config
+	s.taskCfgs[clone.Name] = taskCfg
+
+	return nil
+}
+
+// RemoveTaskCfg removes the config of task
+func (s *Scheduler) RemoveTaskCfg(task string) error {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	// do not remove task if has subtask
+	if _, ok := s.subTaskCfgs[task]; ok {
+		return nil
```

Comment:
or return some error?

Reply:
We have command `stop-task -s source` for sub-task. So actually it is not a error.